dali/kernels/common/join/tensor_join_cpu.h

```diff
@@ -0,0 +1,209 @@
+                   span<const TensorView<StorageCPU, const T>> inputs,
+                   int axis) {
+  SmallVector<int64_t, 8> copy_sizes;
+  for (int t = 0; t < inputs.size(); t++) {
+    copy_sizes[t] = volume(inputs[t].shape.begin() + axis, inputs[t].shape.end());
+  }
+  auto nouter = volume(output.shape.begin(), output.shape.begin() + axis);
+  auto *out = output.data;
+  for (ptrdiff_t outer = 0; outer < nouter; outer++) {
+    for (int t = 0; t < inputs.size(); t++) {
+      auto *src = inputs[t].data + outer * copy_sizes[t];
+      for (ptrdiff_t inner = 0; inner < copy_sizes[t]; inner++) {
+        *out++ = src[inner];
+      }
+    }
+  }
+}
+
+
+template<bool new_axis = true>
+TensorShape<> DetermineShape(span<const TensorShape<>> in_shapes, int axis) {
+  TensorShape<> ret;
+  auto &insh = in_shapes[0];
+  ret.resize(insh.size() + 1);
+  for (int i = 0; i < axis; i++) {
+    ret[i] = insh[i];
+  }
+  ret[axis] = in_shapes.size();
+  for (int i = axis + 1; i < ret.size(); i++) {
+    ret[i] = insh[i - 1];
+  }
+  return ret;
+}
+
+
+template<>
+TensorShape<> DetermineShape<false>(span<const TensorShape<>> in_shapes, int axis) {
+  TensorShape<> ret = in_shapes[0];
+  ret[axis] = 0;
+  for (const auto &sh : in_shapes) {
+    ret[axis] += sh[axis];
+  }
+  return ret;
+}
+
+
+/**
+ * Joins multiple input tensors into one output tensor, along given axis.
+ *
+ * The kernel works in 2 modes: STACK and CONCAT. In CONCAT mode, kernel creates a new tensor,
+ * with joined values along given dimension, e.g.
+ *
+ * arr0 = [[1, 2, 4, 2], [1, 1, 7, 6], [6, 8, 8, 4]])
+ * shape = (3, 4)
+ *
+ * arr1 = [[3, 8, 8, 6], [8, 1, 5, 7], [6, 2, 7, 5]]
+ * shape = (3, 4)
+ *
+ * concatenate([arr0, arr1], axis=1) -> [[1, 2, 4, 2, 3, 8, 8, 6],
+ *                                       [1, 1, 7, 6, 8, 1, 5, 7],
+ *                                       [6, 8, 8, 4, 6, 2, 7, 5]]
+ * shape = (3, 8)
+ *
+ * Stacking, on the other hand, creates new tensor with added dimension, where `axis` points.
+ *
+ * stack([arr0, arr1], axis=1) -> [[[1, 2, 4, 2],
+ *                                  [3, 8, 8, 6]],
+ *
+ *                                 [[1, 1, 7, 6],
+ *                                  [8, 1, 5, 7]],
+ *
+ *                                 [[6, 8, 8, 4],
+ *                                  [6, 2, 7, 5]]]
+ * shape = (3, 2, 4)
+ *
+ * @tparam new_axis if true, STACK mode is applied.
+ */
+template<typename T, bool new_axis = true, int dims = -1>
+struct TensorJoinCpu {
+  ///@{
+  /**
+   * @param in_shapes Shapes of input tensors.
+   * @param axis Axis, along which tensors will be joined. Accepts positive and negative values
+   *             (e.g. axis: -1 <=> ndims-1)
+   */
+  KernelRequirements Setup(KernelContext &ctx, span<const TensorShape<dims>> in_shapes, int axis) {
+    n_input_tensors_ = in_shapes.size();
+    auto ndims = in_shapes[0].sample_dim();
+    DALI_ENFORCE(axis >= -ndims + !new_axis && axis <= ndims - !new_axis,
```

Comment:
As discussed in slack. I believe that this kind of -negative indexing logic doesn't belong to the kernel layer. I'd make the kernel accept:
[0, ndims] for STACK
[0, ndims) for CONCAT
And leave negative indexing to the operator

Reply:
Done